### PR TITLE
Make standard GHC definitions available to hsc2hs

### LINKED ISSUE #107: Error building old-time (due to hsc)
This is the hsc2hs invocation done:
```sh
external/my_ghc/bin/hsc2hs external/haskell_old_time/System/Time.hsc -o /tmp/x.hs -Iexternal/haskell_old_time/include -Ibazel-out/k8-fastbuild/genfiles/external/haskell_old_time/include
```

old-time doesn't include HsTime.h unless it's on GHC: https://github.com/haskell/old-time/blob/master/System/Time.hsc#L110. I can manually run the command with `-D__GLASGOW_HASKELL__` and it works.

```python
load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_library")

genrule(
    name = "config",
    srcs = [
        "configure",
        "install-sh",
        "config.sub",
        "config.guess",
        "include/HsTime.h",
        "include/HsTimeConfig.h.in",
    ],
    outs = ["include/HsTimeConfig.h"],
    cmd = """
        cd `dirname $(location configure)`
        ./configure
        DIR=`pwd`
        cd -
        cp $$DIR/include/HsTimeConfig.h $@
    """,
)

cc_library(
    name = "cbits",
    srcs = [
        "include/HsTime.h",
        ":config",
    ],
    includes = ["include"],
)

haskell_library(
    name = "old-time",
    srcs = glob(["System/Time.hsc"]),
    prebuilt_dependencies = ["base"],
    visibility = ["//visibility:public"],
    deps = [":cbits"],
)
```

### PR DESCRIPTION
**Pull request by mrkkrp**
Close #107.

### CODE REVIEW

**Comment by Fuuzetsu** [comment]
I don't know if this is the right solution. This is a standard GHC-defined macro that should be available. Further, GHC defines more of these including ones for its version and package versions which we can't really replicate easily.

**Comment by mrkkrp** [comment]
Any idea what is the right solution here?

**Comment by Fuuzetsu** [comment]
Nope… I would look at what Cabal does when invoking the tool and why we don't have the same macros available.

File: haskell/hsc2hs.bzl [hunk 1]
```diff
@@ -49,6 +49,8 @@ def _process_hsc_file(ctx, hsc_file):
   hdrs, include_args = cc_headers(ctx)
   args.add(include_args)
 
+  args.add("-D__GLASGOW_HASKELL__")
```
Comment (iphydf) [inline haskell/hsc2hs.bzl]: This needs to be the actual value, e.g. 701, not just be defined.

File: haskell/hsc2hs.bzl [hunk 2]
```diff
@@ -49,6 +49,14 @@ def _process_hsc_file(ctx, hsc_file):
   hdrs, include_args = cc_headers(ctx)
   args.add(include_args)
 
+  # One of the standard definitions that should be available, we'll add them
```
Comment (mboes) [inline haskell/hsc2hs.bzl]: This comment should be updated to point to `platformDefines` in `Cabal/Distribution/Simple/PreProcess.hs`.

File: haskell/hsc2hs.bzl [hunk 3]
```diff
@@ -49,6 +49,14 @@ def _process_hsc_file(ctx, hsc_file):
...
+  # as we discover problematic projects. It is not certain where e.g. Cabal
+  # takes them from, probably it just figures out them one by one, like we
+  # do here.
+  ghc_ver = ctx.toolchains["@io_tweag_rules_haskell//haskell:toolchain"].version
+  ghc_ver = (ghc_ver.replace(".", "") + "000")[:3]
```
Comment (mboes) [inline haskell/hsc2hs.bzl]: Is this conversion correct in all cases? Document the assumptions.

File: haskell/hsc2hs.bzl [hunk 4]
```diff
@@ -17,19 +19,61 @@ def hsc_to_hs(ctx):
...
+    inputs = depset([ghc_defs_dump0]),
+    outputs = [ghc_defs_dump],
+    command = "tail -n +2 $1 > $2",
+    arguments = [
+      ghc_defs_dump0.path,
+      ghc_defs_dump.path,
+    ],
+  )
+
+  return ghc_defs_dump
```
Comment (iphydf) [inline haskell/hsc2hs.bzl]: This probably needs some filtering:

```
bazel-out/k8-fastbuild/bin/external/haskell_old_time/ghc-defs-dump-old-time-1.0.0.h:17:0: warning: "__STDC_HOSTED__" redefined
 #define __STDC_HOSTED__ 1
```

And more importantly, all kinds of things break. I haven't investigated why yet, but I guess some of these defines shouldn't be defined:
```
In file included from /usr/include/x86_64-linux-gnu/sys/time.h:25:0,
                 from external/haskell_old_time/include/HsTime.h:19,
                 from Time.hsc:111:
/usr/include/time.h:75:1: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'typedef'
 typedef __time_t time_t;
 ^~~~~~~
/usr/include/time.h:76:1: error: unknown type name '__END_NAMESPACE_STD'
 __END_NAMESPACE_STD
 ^~~~~~~~~~~~~~~~~~~
```

**Comment by mrkkrp** [comment]
@mboes @iphydf Please take a look, I think now it's OK.

**Pull request merged**